jwt/jwt.go

```diff
@@ -0,0 +1,256 @@
+			}
+			*claims.NotBefore = jwt.NumericDate(int64(*claims.Expiry) - int64(leeway))
+		}
+	}
+
+	// Set clock skew leeway to apply when validating all time-related claims
+	cksLeeway := expected.ClockSkewLeeway
+	if expected.ClockSkewLeeway.Seconds() < 0 {
+		cksLeeway = 0
+	} else if expected.ClockSkewLeeway.Seconds() == 0 {
+		cksLeeway = jwt.DefaultLeeway
+	}
+
+	// Validate claims by asserting they're as expected
+	if expected.Issuer != "" && expected.Issuer != claims.Issuer {
+		return nil, fmt.Errorf("invalid issuer (iss) claim")
+	}
+	if expected.Subject != "" && expected.Subject != claims.Subject {
+		return nil, fmt.Errorf("invalid subject (sub) claim")
+	}
+	if expected.ID != "" && expected.ID != claims.ID {
+		return nil, fmt.Errorf("invalid ID (jti) claim")
+	}
+	if err := validateAudience(expected.Audiences, claims.Audience); err != nil {
+		return nil, fmt.Errorf("invalid audience (aud) claim: %w", err)
+	}
+
+	// Validate that the token is not expired with respect to the current time
+	now := time.Now()
+	if expected.Now != nil {
+		now = expected.Now()
+	}
+	if claims.NotBefore != nil && now.Add(cksLeeway).Before(claims.NotBefore.Time()) {
+		return nil, errors.New("invalid not before (nbf) claim: token not yet valid")
+	}
+	if claims.Expiry != nil && now.Add(-cksLeeway).After(claims.Expiry.Time()) {
+		return nil, errors.New("invalid expiration time (exp) claim: token is expired")
+	}
+	if claims.IssuedAt != nil && now.Add(cksLeeway).Before(claims.IssuedAt.Time()) {
+		return nil, errors.New("invalid issued at (iat) claim: token issued in the future")
+	}
+
+	return allClaims, nil
+}
+
+// validateSigningAlgorithm checks whether the JWS "alg" (Algorithm) header
+// parameter value for the given JWT matches any given in expectedAlgorithms.
+// If expectedAlgorithms is empty, "RS256" will be expected by default.
+func validateSigningAlgorithm(token string, expectedAlgorithms []string) error {
+	jws, err := jose.ParseSigned(token)
+	if err != nil {
+		return err
+	}
+
+	switch len(jws.Signatures) {
+	case 0:
+		return fmt.Errorf("token must be signed")
+	case 1:
+	default:
+		return fmt.Errorf("token with multiple signatures not supported")
+	}
+
+	if len(expectedAlgorithms) == 0 {
+		expectedAlgorithms = []string{string(jose.RS256)}
+	}
+
+	alg := jws.Signatures[0].Header.Algorithm
+	if !contains(expectedAlgorithms, alg) {
+		return fmt.Errorf("token signed with unexpected algorithm")
+	}
+
+	return nil
+}
+
+// validateAudience returns an error if audClaim does not contain any audiences
+// given by expectedAudiences. If expectedAudiences is empty, it skips validation
+// and returns nil.
+func validateAudience(expectedAudiences, audClaim []string) error {
+	if len(expectedAudiences) == 0 {
+		return nil
+	}
+
+	for _, v := range expectedAudiences {
+		if contains(audClaim, v) {
+			return nil
+		}
+	}
+
+	return errors.New("audience claim does not match any expected audience")
+}
+
+func contains(sl []string, st string) bool {
```

Comment:
This looks exactly like `StrListContains` from `oidc/internal/strutils`. @jimlambrt thoughts on moving `oidc/internal` into the top-level `cap` directory?

Reply:
Hum... it's such a small simple bit of behavior, I'm okay with it being copied.